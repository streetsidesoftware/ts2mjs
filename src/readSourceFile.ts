import { readFile } from './fileUtils.js';
import type { SourceFile, SourceMap } from './SourceFile.js';
import { fileURLToPath, pathToFileURL } from 'url';

export const SOURCE_MAP_URL_MARKER = '//' + '# ' + 'sourceMappingURL' + '=';

export interface SourceMapRef {
    index: number;
    url: URL;
}

export async function readSourceFile(filename: string): Promise<SourceFile> {
    const content = await readFile(filename);
    const srcMapRef = getSourceMapRef(filename, content);
    const sourceMap = srcMapRef && (await readSourceMap(srcMapRef.url));

    const srcContent = (srcMapRef && content.slice(0, srcMapRef.index)) || content;

    return {
        srcFilename: filename,
        content: srcContent,
        map: sourceMap,
    };
}

export function getSourceMapRef(filename: string, content: string): SourceMapRef | undefined {
    const srcMapIdx = content.lastIndexOf(SOURCE_MAP_URL_MARKER);
    if (srcMapIdx < 0) return undefined;
    const srcMapUrl = content.slice(srcMapIdx + SOURCE_MAP_URL_MARKER.length).trim();

    try {
        const url = new URL(srcMapUrl, pathToFileURL(filename));
        return { index: srcMapIdx, url };
    } catch (e) {
        return undefined;
    }
}

export async function readSourceMap(url: URL): Promise<SourceMap | undefined> {
    try {
        return { map: await readFile(fileURLToPath(url)), url };
    } catch (e) {
        // do not break if source map file is missing.
        return undefined;
    }
}
