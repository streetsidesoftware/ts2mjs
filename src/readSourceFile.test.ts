import { pathToFileURL } from 'url';
import { describe, expect, test } from 'vitest';

import { getSourceMapRef, readSourceFile, readSourceMap, SOURCE_MAP_URL_MARKER } from './readSourceFile.js';
import { resolveFixture as ff } from './test.util.js';

const oc = (obj: unknown) => expect.objectContaining(obj);

describe('ReadSourceFile', () => {
    test.each`
        file
        ${'sample/lib/index.js.map'}
        ${'sample/lib/notfound.js.map'}
    `('readSourceMap $file', async ({ file }) => {
        file = ff(file);
        const url = pathToFileURL(file);
        expect((await readSourceMap(url))?.map).toMatchSnapshot();
    });

    test.each`
        file
        ${'sample/lib/index.js'}
    `('readSourceFile $file', async ({ file }) => {
        file = ff(file);
        expect((await readSourceFile(file))?.map?.map).toMatchSnapshot();
    });

    test.each`
        file                        | expected
        ${'sample/lib/notfound.js'} | ${oc({ code: 'ENOENT', message: expect.stringContaining('no such file') })}
    `('readSourceFile error $file', async ({ file, expected }) => {
        file = ff(file);
        await expect(readSourceFile(file)).rejects.toEqual(expected);
    });

    test.each`
        file                        | content                                                        | expected
        ${'sample/lib/index.js'}    | ${`export const x = 5;\n${SOURCE_MAP_URL_MARKER}index.js.map`} | ${oc({ index: 20 })}
        ${'sample/lib/notfound.js'} | ${`\n${SOURCE_MAP_URL_MARKER}notfound.js.map`}                 | ${oc({ index: 1 })}
        ${'sample/lib/url.js'}      | ${`\n${SOURCE_MAP_URL_MARKER}https::url.js.map`}               | ${undefined}
    `('getSourceMapRef $file', async ({ file, content, expected }) => {
        file = ff(file);
        expect(await getSourceMapRef(file, content)).toEqual(expected);
    });
});
