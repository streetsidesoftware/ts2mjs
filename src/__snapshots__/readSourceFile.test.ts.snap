// Vitest Snapshot v1

exports[`ReadSourceFile > readSourceFile 'sample/lib/index.js' 1`] = `"{\\"version\\":3,\\"file\\":\\"index.js\\",\\"sourceRoot\\":\\"\\",\\"sources\\":[\\"../src/index.ts\\"],\\"names\\":[],\\"mappings\\":\\"AAaA,OAAO,EAAE,YAAY,EAAE,MAAM,aAAa,CAAC;AAE3C,OAAO,EAAE,WAAW,EAAE,MAAM,qBAAqB,CAAC\\"}"`;

exports[`ReadSourceFile > readSourceMap 'sample/lib/index.js.map' 1`] = `"{\\"version\\":3,\\"file\\":\\"index.js\\",\\"sourceRoot\\":\\"\\",\\"sources\\":[\\"../src/index.ts\\"],\\"names\\":[],\\"mappings\\":\\"AAaA,OAAO,EAAE,YAAY,EAAE,MAAM,aAAa,CAAC;AAE3C,OAAO,EAAE,WAAW,EAAE,MAAM,qBAAqB,CAAC\\"}"`;

exports[`ReadSourceFile > readSourceMap 'sample/lib/notfound.js.map' 1`] = `undefined`;
